test/gherkin.go

```diff
@@ -12,195 +14,161 @@ const (
+func Given(description string, setup func()) GivenStatement {
+	return GivenStatement{
 		label: []string{_given, description},
-		test:  []func(t *testing.T){setup},
+		test:  setup,
 	}
 }
 
 // When is an independent trigger that can be used to start a statement in a Branch
-func When(description string, setup func(t *testing.T)) ThenStatement {
-	return then{
+func When(description string, setup func()) WhenStatement {
+	return WhenStatement{
 		label: []string{_when, description},
-		test:  []func(t *testing.T){setup},
+		test:  setup,
 	}
 }
 
 // Then is an independent outcome check that can be used to start a statement in a Branch
-func Then(description string, setup func(t *testing.T)) Runner {
-	return runner{
+func Then(description string, setup func(t *testing.T)) ThenStatements {
+	return []thenStatement{{
 		label: []string{_then, description},
-		test:  []func(t *testing.T){setup},
-	}
+		test:  setup,
+	}}
 }
 
-// TestCases takes an array of arbitrary items to then generate a test for each of the test case items
-func TestCases[T any](items []T) ForEachStatement[T] {
-	return testCases[T]{
-		items: items,
+// Given adds an additional precondition
+func (g GivenStatement) Given(description string, setup func()) GivenStatement {
+	return GivenStatement{
+		label: append(g.label, _and, _given, description),
+		test:  func() { g.test(); setup() },
 	}
 }
 
-// Given adds an additional precondition
-func (g given) Given(description string, setup func(t *testing.T)) WhenStatement {
-	return when{
-		label: append(g.label, _given, description),
-		test:  append(g.test, setup),
+func (g GivenStatement) Given2(g2 GivenStatement) GivenStatement {
+	return GivenStatement{
+		label: mergeLabels(g.label, g2.label),
+		test:  func() { g.test(); g2.test() },
 	}
 }
 
-// And allows to concatenate an additional precondition
-func (w when) And() GivenStatement {
-	return given{
-		label: append(w.label, _and),
-		test:  w.test,
+// When adds a trigger to the test path
+func (g GivenStatement) When(description string, setup func()) WhenStatement {
+	return WhenStatement{
+		label: append(g.label, _when, description),
+		test:  func() { g.test(); setup() },
 	}
 }
 
-// When adds a trigger to the test path
-func (w when) When(description string, setup func(t *testing.T)) ThenStatement {
-	return then{
-		label: append(w.label, _when, description),
-		test:  append(w.test, setup),
+func (g GivenStatement) When2(w WhenStatement) WhenStatement {
+	return WhenStatement{
+		label: mergeLabels(g.label, w.label),
+		test:  func() { g.test(); w.test() },
 	}
 }
 
 // Branch allows test branching by adding multiple sub-statements after a Given
-func (w when) Branch(tests ...Runner) Runner {
-	checkFirstWordOfLabels(tests, assertNotTHEN)
-	return branch(w.label, w.test, tests)
-}
+func (g GivenStatement) Branch(thens ...ThenStatements) ThenStatements {
+	slices.ForEach(thens, func(thens ThenStatements) { checkFirstWordOfLabels(thens, assertNotTHEN) })
```

Comment:
This is unrelated, but while using `slices.ForEach`, I find it hard to use it since it's hard to handle error if the "callback" function has to return one.